syntax = "proto3";

import "google/protobuf/timestamp.proto";
import "sensr_proto/global_type.proto";
import "sensr_proto/type.proto";

package sensr_proto;

message ZoneEvent {
  enum Type {
    NONE = 0;
    ENTRY = 1;
    EXIT = 2;
    LOITERING = 3;
    EXCEEDS_SPEED = 4;
  }
  message Object {
    int32 id = 1;
    Vector3 position = 2;
    float heading = 3;
    Vector3 velocity = 4;
  }
  google.protobuf.Timestamp timestamp = 1;
  int32 id = 2;
  Type type = 3;
  Object object = 4;
}

message LosingEvent {
  google.protobuf.Timestamp timestamp = 1;
  int32 id = 2;
  Vector3 position = 3;
  float heading = 4;
}

message SystemHealth {
  enum Status {
    NONE = 0;
    OK = 1;
    STORAGE_SHORTAGE = 2;
    SLOWDOWN = 3;
    INTERNAL_ERROR = 4;
  }  
  message Node {
    enum Status {
      NONE = 0;
      OK = 1;
      ROS_ERROR = 2;
      LOST_CONNECTION = 3;
    }
    enum SensorStatus {
      SENSOR_DEAD = 0;
      SENSOR_ALIVE = 1;
    }
    Status status = 1;
    map<string, SensorStatus> sensors = 3;
  }
  Status master = 1;
  map<string, Node> nodes = 2;
}

message ZoneConfig {
  int32 id = 1;
  string name = 2;
  PolygonBox pbox = 3;
  ZoneType type = 4;
}

message StreamMessage {
  repeated Object objects = 1;
  repeated ZoneConfig zones = 101;
  SystemHealth health = 102;
}

message EventMessage {
  repeated ZoneEvent zone = 1;
  repeated LosingEvent losing = 2;
  SystemHealth health = 3;
}

message CustomMessage {
  repeated PolygonBox field_of_regard = 1;
  repeated PredictedReachableSet predicted_reachable_set = 2;
}

// Actual message using in SENSR SDK
message OutputMessage {
  google.protobuf.Timestamp timestamp = 1;
  reserved 2, 101, 102, 103; // framework 1.0 used these.
  StreamMessage stream = 3;
  EventMessage event = 201;
  CustomMessage custom = 301;
}
